### ISSUE TITLE
Fix output from update_translations.py

### ISSUE THREAD

**Issue by aaronchantrill** [NaomiProject/Naomi#214]
## Detailed Description

Naomi uses gettext to provide translations into other languages. We use the update_translations.py script to extract raw strings from naomi core and core plugins and update the locale folders with <language>.pot files, then merge them with the existing .po files. Unfortunately, this introduces a lot of duplicates and errors into the resulting .po file. What would be good would be to use regex to analyze the new .pot and old .po files manually and import matching translations and also copy relevant information from the headers.

## Context

This program was written to allow developers to easily and quickly regenerate .po language translation files. The Naomi project always used gettext to allow multiple translations of text strings to be created and included a requirement that compile_translations.sh be used to generate the .mo files for your system after downloading, but didn't include much information on how developers were to generate these files in the first place. Generating and updating translation files required running pygettext.py on each folder every time a string literal was altered. This made it a big pain to update the translations or add a new language.

While this program automated the process of generating the language files, it is far from perfect. When creating a new .pot file, it generates a msgstr header with default information, then the msgcat program combines the headers one after the other, which causes problems for gettext. So after running "update_translations.py" the developer has to manually go through all the files and correct all the headers. For instance, the header for the old de-DE.po file might be
```
# #-#-#-#-#  de-DE.po (Naomi 2.2-dev)  #-#-#-#-#
# Naomi core translations
# Copyright (C) 2015 Shubhro Saha, Charlie Marsh & Jan Holthuis
# Portions copyright (C) 2018 others
# This file is distributed under the same license as the Naomi package.
# Jan Holthuis <holthuis.jan@googlemail>, 2015.
#
#, fuzzy
msgid ""
msgstr ""
"#-#-#-#-#  de-DE.po (Naomi 2.2-dev)  #-#-#-#-#\n"
"Project-Id-Version: Naomi 2.2-dev\n"
"Language: de_DE\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"POT-Creation-Date: 2018-10-28 22:49+EDT\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: ENCODING\n"
"Generated-By: pygettext.py 1.5\n"
```
and the newly generated .pot file would be
```
# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR ORGANIZATION
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
msgid ""
msgstr ""
"Project-Id-Version: PACKAGE VERSION\n"
"POT-Creation-Date: 2019-10-04 14:58-0400\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Generated-By: pygettext.py 1.5\n"
```
and the msgcat combined file would be
```
# #-#-#-#-#  de-DE.po (Naomi 2.2-dev)  #-#-#-#-#
# #-#-#-#-#  de-DE.po (Naomi 2.2-dev)  #-#-#-#-#
# Naomi core translations
# Copyright (C) 2015 Shubhro Saha, Charlie Marsh & Jan Holthuis
# Portions copyright (C) 2018 others
# This file is distributed under the same license as the Naomi package.
# Jan Holthuis <holthuis.jan@googlemail>, 2015.
#
# #-#-#-#-#  temp.pot (PACKAGE VERSION)  #-#-#-#-#
# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR ORGANIZATION
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
#, fuzzy
msgid ""
msgstr ""
"#-#-#-#-#  de-DE.po (Naomi 2.2-dev)  #-#-#-#-#\n"
"#-#-#-#-#  de-DE.po (Naomi 2.2-dev)  #-#-#-#-#\n"
"Project-Id-Version: Naomi 2.2-dev\n"
"Language: de_DE\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"POT-Creation-Date: 2018-10-28 22:49+EDT\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Generated-By: pygettext.py 1.5\n"
"#-#-#-#-#  temp.pot (PACKAGE VERSION)  #-#-#-#-#\n"
"Project-Id-Version: PACKAGE VERSION\n"
"POT-Creation-Date: 2019-10-04 14:58-0400\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Generated-By: pygettext.py 1.5\n"
```
This requires the user to delete the duplicate lines while preserving the old comments and deleting the new comments, then replacing the old "POT-Creation-Date", "PO-Revision-Date", "X-Generator" and "Generated-By" information while preserving important lines like "#, fuzzy", "Plural-Forms", "Language" and "Project-Id-Version".

Also, since msgcat is unaware of the contents of the source document, its output includes lines that no longer appear in the source, and lines that have moved from one location to another in the source file now have multiple unversioned or dated comments, both of which combine to create a confusing, less readable file.

Finally, the update_translations.py file is still in python2.7 and needs to be updated to python3 so we aren't requiring two different versions of python just for this project.

## Possible Implementation

A lot of this could be accomplished by looping over the generated .pot file and replacing or adding lines from the original header file, or using regex to parse certain information from the new .pot file and inserting it into a standard header.

It might also be nice to have a small sqlite database of translations, so that the script could read through the file and insert the translations it knows into the .pot file instead of using msgcat to merge them.

## Your Environment

* Version used: naomi-dev
* Environment name and version (e.g. PHP 5.4 on nginx 1.9.1): Python 3.7
* Server type and version: Raspberry Pi 3B+
* Operating System and version: Raspbian Buster

**Comment by kaozdl** [comment]
i can take a look at this

**Comment by aaronchantrill** [comment]
This was fixed in pull request #248 https://github.com/NaomiProject/Naomi/pull/248
